core/src/main/scalajvm/scalapb/zio_grpc/server/ZServerCallHandler.scala

```diff
@@ -108,30 +111,62 @@ object ZServerCallHandler {
       call: ZServerCall[Res],
       stream: ZStream[Any, Status, Res]
   ): ZIO[Any, Status, Unit] = {
-    def innerLoop(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Boolean] =
-      queue.take
-        .flatMap {
-          case Exit.Success(res)   => call.sendMessage(res).as(true)
-          case Exit.Failure(cause) =>
-            cause.failureOrCause match {
-              case Left(Some(status)) => ZIO.fail(status)
-              case Left(None)         => ZIO.succeed(false)
-              case Right(cause)       => ZIO.failCause(cause)
+    def takeFromQueue(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Unit] =
+      queue.takeAll.flatMap(takeFromCache(_, queue))
+
+    def takeFromCache(
+        xs: Chunk[Exit[Option[Status], Res]],
+        queue: Dequeue[Exit[Option[Status], Res]]
+    ): ZIO[Any, Status, Unit] =
+      ZIO.suspendSucceed {
+        @tailrec def innerLoop(loop: Boolean, i: Int): IO[Status, Unit] =
+          if (loop) {
+            xs(i) match {
+              case Failure(cause) =>
+                cause.failureOrCause match {
+                  case Left(Some(status)) =>
+                    ZIO.fail(status)
+                  case Left(None)         =>
+                    ZIO.unit
+                  case Right(cause)       =>
+                    ZIO.failCause(cause)
+                }
+              case Success(value) =>
+                call.call.sendMessage(value)
+                // at this point we have consumed i + 1 elements from the chunk
+                val newI = i + 1
+                // the loop iteration may only continue if the call can
+                // still accept elements and we have more elements to send
+                innerLoop(call.call.isReady && newI < xs.length, newI)
             }
-        }
-        .repeatWhileZIO(res => call.isReady.map(_ && res))
+          } else if (call.call.isReady)
+            // ^ if we reached the end of the chunk but the call can still
+            // proceed, we pull from the queue and continue
+            takeFromQueue(queue)
```

Comment:
Is there a possibility for unbounded recursion here?

Reply:
In terms of revision, no, as `i` cannot be negative. But if the stream is infinite, the effect will run ad infinitum. But that is the case already.

Reply:
Oh boy, typing on the train with too many people around makes me write nonsense. The stream part is correct, but I've inlined the index check as it feels easier to get right there.